File: Baekjoon/yeongho/Week_2/4_BJ1167.cs
class Program
{
    public static void Main()
    {
        BJ1167 simulation = new BJ1167();
        
        simulation.Initialize();
        simulation.StartSimulation();
    }
}

public class BJ1167
{
    private Node2 bestFarNode;
    private int distanceValue;
    private Dictionary<int, Node2> nodeDic;

    public void Initialize()
    {
        int nodeCnt = int.Parse(Console.ReadLine());
        nodeDic = new Dictionary<int, Node2>();
        for (int i = 1; i < nodeCnt + 1; i++)
        {
            nodeDic.Add(i, new Node2(i));
        }

        for (int i = 1; i < nodeCnt + 1; i++)
        {
            var dimensions = Array.ConvertAll(Console.ReadLine().Split(), int.Parse);
            int j = 1;
            var nodeIdx = dimensions[0];
            do
            {
                nodeDic[nodeIdx].AddNodeLinked(dimensions[j], dimensions[j + 1]);
                j += 2;
            } while (dimensions[j] != -1);
        }
    }
    
    public void StartSimulation()
    {
        // 1. 노드부터 시작하여 가장 깊이있는 노드를 찾는다.
        BFS(nodeDic[1]);

        ResetNode();
        
        // 2. 가장 깊이 있는 노드부터 시작해서 거꾸러 찾아 올라간다.
        BFS(bestFarNode);

        Console.WriteLine(distanceValue);
    }

    private void BFS(Node2 node)
    {
        Queue<Node2> qList = new Queue<Node2>();
        qList.Enqueue(node);
        node.isVisited = true;

        while (qList.Count > 0)
        {
            Node2 curNode = qList.Dequeue();
            foreach (var key in curNode.linkedNodeList.Keys)
            {
                if (nodeDic[key].isVisited) continue;
                
                nodeDic[key].isVisited = true;
                nodeDic[key].accumulatedSum = curNode.accumulatedSum + curNode.linkedNodeList[key];
                if (distanceValue < nodeDic[key].accumulatedSum)
                {
                    distanceValue = nodeDic[key].accumulatedSum;
                    bestFarNode = nodeDic[key];
                }
                
                qList.Enqueue(nodeDic[key]);
            }
        }
    }

    private void ResetNode()
    {
        foreach (var nodeDicValue in nodeDic.Values)
        {
            nodeDicValue.isVisited = false;
            nodeDicValue.accumulatedSum = 0;
        }
    }
    
    class Node2
    {
        public int index;           //노드의 인덱스
        public int accumulatedSum; // 누적합
        public bool isVisited;      // 방문 유무
        public Dictionary<int, int> linkedNodeList; // key : idx, value : distance

        public Node2(int idx)
        {
            index = idx;
            accumulatedSum = 0;
            isVisited = false;
            linkedNodeList = new Dictionary<int, int>();
        }

        public void AddNodeLinked(int node, int dist)
        {
            linkedNodeList.Add(node, dist);
        }
    }
    
}

File: Baekjoon/yeongho/Week_2/4_BJ1167.py
from collections import deque, defaultdict

class Node:
    def __init__(self, index):
        self.index = index
        self.accumulated_sum = 0
        self.is_visited = False
        self.linked_nodes = {}  # key: node index, value: distance

    def add_linked_node(self, node, dist):
        self.linked_nodes[node] = dist

class BJ1167:
    def __init__(self):
        self.best_far_node = None
        self.distance_value = 0
        self.node_dict = {}

    def initialize(self):
        node_count = int(input())
        self.node_dict = {i: Node(i) for i in range(1, node_count + 1)}

        for _ in range(node_count):
            dimensions = list(map(int, input().split()))
            node_idx = dimensions[0]
            j = 1
            while dimensions[j] != -1:
                self.node_dict[node_idx].add_linked_node(dimensions[j], dimensions[j + 1])
                j += 2

    def start_simulation(self):
        # 1. Start BFS from node 1 to find the farthest node
        self.bfs(self.node_dict[1])

        # Reset node states
        self.reset_nodes()

        # 2. Start BFS from the farthest node to calculate the diameter
        self.bfs(self.best_far_node)

        print(self.distance_value)

    def bfs(self, node):
        queue = deque([node])
        node.is_visited = True

        while queue:
            current_node = queue.popleft()

            for key, distance in current_node.linked_nodes.items():
                next_node = self.node_dict[key]
                if next_node.is_visited:
                    continue

                next_node.is_visited = True
                next_node.accumulated_sum = current_node.accumulated_sum + distance

                if self.distance_value < next_node.accumulated_sum:
                    self.distance_value = next_node.accumulated_sum
                    self.best_far_node = next_node

                queue.append(next_node)

    def reset_nodes(self):
        for node in self.node_dict.values():
            node.is_visited = False
            node.accumulated_sum = 0

if __name__ == "__main__":
    simulation = BJ1167()
    simulation.initialize()
    simulation.start_simulation()
